### PR TITLE
Ck dm shopping list order

### LINKED ISSUE #12: 12. As a user, I want to view a list of my shopping list items in order of how soon I am likely to need to buy each of them again so that it’s clear what I need to buy soon.
Possible item states are as follows:
* Need to buy soon (fewer than 7 days)
* Need to buy kind of soon (between 7 & 30 days)
* Need to buy not soon (more than 30 days)
* Inactive (when there’s only 1 purchase in the database or the purchase is really out of date [the time that has elapsed since the last purchase is 2x what was estimated])

AC:
* Items in the list are shown as visually distinct (e.g., with a different background color on the list item) according to how soon the item is expected to be bought again: Soon, Kind of soon, Not soon, Inactive
* Items should be sorted by the estimated number of days until the next purchase
* Items with the same number of estimated days until next purchase should be sorted alphabetically
* Items in the different states should be described distinctly when read by a screen reader
 
Notes:
* The criteria for an item being inactive has tripped up teams in the past. It's been re-worded to be more clear, but the gist is that an item is inactive if the time that has elapsed since its last purchase is twice (or more) what was estimated the last time it was purchased.
* E.g., if the last estimate was 14 days and 28 days have elapsed, the item is considered inactive.
* Teams in the past have also overengineered the solution here by breaking the list out into separate groups, but it should be achievable in a single list where the visual treatment is different based on what category the item falls into.
* Be sure to not only use color to denote how soon the user will need to buy the item, as noted in the wireframe. https://github.com/the-collab-lab/smart-shopping-list/blob/main/_resources/wireframes/list-populated.png
* Good luck, you got this!

### PR DESCRIPTION
**Pull request by diyanamendoza**
## Description

This PR adds an item sort to the useFirebaseSnapshot custom hook in order to return a list of items arranged based on next purchase date. It also adds an itemStatus utility function that returns "soon," "kind of soon," "not soon," or "inactive" also based on next purchase date. The itemStatus utility is used in the className of the individual item render in order to be styled differently based status, as well as in the aria label to provide a screen-reader-friendly status message. 

Please note: the current status-based styling is meant to be a placeholder ahead of actual design implementation.

## Related Issue

Closes #12 

## Acceptance Criteria

* Items in the list are shown as visually distinct (e.g., with a different background color on the list item) according to how soon the item is expected to be bought again: Soon, Kind of soon, Not soon, Inactive
* Items should be sorted by the estimated number of days until the next purchase
* Items with the same number of estimated days until next purchase should be sorted alphabetically
* Items in the different states should be described distinctly when read by a screen reader

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

| Before | After |
|---|---|
| Screen Shot 2022-02-15 at 5 45 13 PM | Screen Shot 2022-02-15 at 5 45 49 PM |

## Testing Steps / QA Criteria

Pull down branch and run the app. Items in your list should be rendered in different colors based on status, and ordered based on next purchase date. In ItemList.js, you can comment back in lines 161-163 in order to see the actual next purchase/estimated purchase interval data for each item to confirm accurate sort order. Try adding new items to check that those are ordered and rendered as expected. For example, a new item added as "Kind of soon" should show up in orange towards the middle of the list. 

Note: there is also a Jest test that is included that tests the itemStatus utility function used in this PR. You can run it by typing: `npm test` in the command line.

### CODE REVIEW

File: src/components/ItemList.js [hunk 1]
```diff
@@ -113,7 +116,13 @@ const ItemList = () => {
           <ul>
             {filteredResults
               ? filteredResults.map((item) => (
-                  <li key={item.id}>
+                  <li
+                    key={item.id}
+                    aria-label={`Need to buy ${item.data.name} ${itemStatus(
+                      item,
+                    )}`}
+                    className={itemStatus(item).replace(/\s+/g, '')}
```
Comment (cynthiaeddy) [inline src/components/ItemList.js:126]: clever!

**Review by cynthiaeddy — APPROVED** [review]
nicely done!

File: src/components/ItemList.js [hunk 2]
```diff
@@ -133,7 +142,13 @@ const ItemList = () => {
                   </li>
                 ))
               : docs.map((item) => (
-                  <li key={item.id}>
+                  <li
+                    key={item.id}
+                    aria-label={`Need to buy ${item.data.name} ${itemStatus(
```
Comment (k-monnik) [inline src/components/ItemList.js]: the dynamic aria label is awesome!!

File: src/hooks/useFirebaseSnapshot.js [hunk 3]
```diff
@@ -19,6 +19,22 @@ export default function useFirebaseSnapshot() {
...
+          ? (sortNum2 = item2.data['next purchase'])
+          : (sortNum2 = item2.data['estimated purchase interval']);
+        return (
+          //sort with sort nums, or if nums equal, sort based on alphabetical order of item name
+          sortNum1 - sortNum2 || item1.data.name.localeCompare(item2.data.name)
+        );
+      });
       setDocs(items);
       setLoading(false);
     });
```
Comment (k-monnik) [inline src/hooks/useFirebaseSnapshot.js:50]: This is brilliant, love how y'all implemented all the sorting so concisely 🔥🤩

**Review by k-monnik — APPROVED** [review]
Great work, y'all! The sort feature works well, the itemStatus utility is a great addition. My only feedback is more of a nitpick, but was thinking it might make more sense to have the items that are in inactive status render at the bottom of the list instead of the top, so that items to be purchased soonest are at the top of the list. Maybe that is something we can discuss while styling?

File: src/hooks/useFirebaseSnapshot.js [hunk 4]
```diff
@@ -19,6 +19,22 @@ export default function useFirebaseSnapshot() {
       querySnapshot.forEach((doc) => {
         items.push({ data: doc.data(), id: doc.id });
       });
+      //sort items
```
Comment (larsbrekken) [inline src/hooks/useFirebaseSnapshot.js:23]: Question about the sort order: Right now the inactive items are shown first in the list (see below) - is this intentional? I would have expected the inactive items to show up at the bottom, out of the way.

image

File: src/utils/itemStatus.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import { Timestamp } from 'firebase/firestore';
```
Comment (larsbrekken) [inline src/utils/itemStatus.js:1]: This logic is a bit convoluted, so I was excited to see that you have included unit tests for it! 👍

File: src/hooks/useFirebaseSnapshot.js (hunk 4 above)
Comment (larsbrekken) [reply src/hooks/useFirebaseSnapshot.js:23]: I think the sort logic is an example of logic that it would have been nice to see unit tests for. It's pretty hard to follow otherwise. In order to do that you'd probably want to pull the logic into a separate function, and possibly a separate file.

It's not always easy to know how much effort to spend on making your code testable and on adding tests. In this case I think it may have been worth it.

File: src/components/ItemList.js (hunk 1 above)
Comment (larsbrekken) [reply src/components/ItemList.js:126]: I also like the concise nature of it. I have two minor comments about it - nothing you have to anything with, but something to keep in mind:
1. Looks like there is some amount of duplication here, with the `<li>` attributes created both here and farther down. It's not always worth it to spend effort on reducing duplication, and some of it is ok. In this case I think it's right on the edge of whether it is worth creating a reusable function for it or not.
2. I'm not sure if the aria-label will be sensible for inactive items ("Need to buy cheese inactive").

File: src/hooks/useFirebaseSnapshot.js (hunk 4 above)
Comment (larsbrekken) [reply src/hooks/useFirebaseSnapshot.js:23]: It's also worth considering whether sorting items belongs here, in the reusable Firebase snapshot, or if the sorting should be applied only by the users of the snapshot that wants to have sorted items.

**Review by larsbrekken — CHANGES_REQUESTED** [review]
Hi folks, great work on the last remaining piece of functionality!

You'll see that I've added a handful of notes and a few nitpicky comments. Nothing critical at all, but I few things I wanted to bring up for discussion. It's very late in the week by now, and it's not a big deal if you don't get to these in time for our meeting tomorrow, and we can just talk through these then instead. Do what you can!

File: src/components/ItemList.js (hunk 1 above)
Comment (chriskorsak) [reply src/components/ItemList.js:126]: We have fixed the aria-label for inactive items. It should now be more readable.

File: src/hooks/useFirebaseSnapshot.js (hunk 4 above)
Comment (chriskorsak) [reply src/hooks/useFirebaseSnapshot.js:23]: We could not agree more on inactive items needing to be at the bottom of the list....This should now be fixed after assigning a very high sortNum1, sortNum2 value if the item is inactive. As it stands, the sorting still is located in the custom hook, but it is on our refactoring radar.

Comment (larsbrekken) [reply src/hooks/useFirebaseSnapshot.js:23]: The sort order looks great now! 
image

File: src/components/ItemList.js (hunk 1 above)
Comment (larsbrekken) [reply src/components/ItemList.js:126]: Yep, the labels look more sensible now: 
image

**Review by larsbrekken — APPROVED** [review]
